lib/trento/application/integration/discovery/payloads/cloud_discovery_payload.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Trento.Integration.Discovery.CloudDiscoveryPayload do
+  @moduledoc """
+  Cloud discovery integration event payload
+  """
+
+  @required_fields []
+
+  use Trento.Type
+  import PolymorphicEmbed, only: [cast_polymorphic_embed: 3]
+
+  deftype do
+    field :provider, Ecto.Enum, values: [:aws, :gcp, :azure, :unknown], default: :unknown
+
+    field :metadata, PolymorphicEmbed,
+      types: [azure: __MODULE__.AzureMetadata],
+      on_type_not_found: :nilify,
+      on_replace: :update
+  end
+
+  def changeset(event, attrs) do
+    enriched_attrs = enrich_metadata_type(attrs)
+
+    event
+    |> cast(enriched_attrs, [:provider])
+    |> cast_polymorphic_embed(:metadata, required: false)
+  end
+
+  defp enrich_metadata_type(%{"provider" => provider, "metadata" => %{} = metadata} = attrs),
+    do: %{attrs | "metadata" => Map.put(metadata, "__type__", provider)}
+
+  defp enrich_metadata_type(attrs), do: attrs
+
+  defmodule AzureMetadata do
```

Comment:
I think having the `AzureMetadata` in other file would be nicer, as this will expand when we have other providers. But well, we can change overtime of course!

Reply:
I thought that too, in the future for sure we will have the need for separate files, I agree. :+1: